pkg/storage/storage.go

```diff
@@ -80,69 +84,121 @@ func (p *storage) GetContainerMetrics(pods ...apitypes.NamespacedName) ([]api.Ti
 
 	timestamps := make([]api.TimeInfo, len(pods))
 	resMetrics := make([][]metrics.ContainerMetrics, len(pods))
-
-	for i, pod := range pods {
-		metricPoint, present := p.pods[pod]
+	for podIdx, pod := range pods {
+		contPoints, present := p.pods[pod]
 		if !present {
 			continue
 		}
 
-		contMetrics := make([]metrics.ContainerMetrics, len(metricPoint.Containers))
-		var earliestTS *time.Time
-		for i, contPoint := range metricPoint.Containers {
-			contMetrics[i] = metrics.ContainerMetrics{
+		contMetrics := make([]metrics.ContainerMetrics, len(contPoints))
+		var earliestTS time.Time
+		var contIdx int
+		for contName, contPoint := range contPoints {
+			prevContPoint := p.prevPods[pod][contName]
+			cpuUsage := cpuUsageOverTime(contPoint.MetricsPoint, prevContPoint.MetricsPoint)
+			contMetrics[contIdx] = metrics.ContainerMetrics{
 				Name: contPoint.Name,
 				Usage: corev1.ResourceList{
-					corev1.ResourceName(corev1.ResourceCPU):    contPoint.CpuUsage,
+					corev1.ResourceName(corev1.ResourceCPU):    cpuUsage,
 					corev1.ResourceName(corev1.ResourceMemory): contPoint.MemoryUsage,
 				},
 			}
-			if earliestTS == nil || earliestTS.After(contPoint.Timestamp) {
-				ts := contPoint.Timestamp // copy to avoid loop iteration variable issues
-				earliestTS = &ts
+			if earliestTS.IsZero() || earliestTS.After(contPoint.Timestamp) {
+				earliestTS = contPoint.Timestamp
 			}
+			contIdx++
 		}
-		if earliestTS == nil {
-			// we had no containers
-			earliestTS = &time.Time{}
-		}
-		timestamps[i] = api.TimeInfo{
-			Timestamp: *earliestTS,
+		resMetrics[podIdx] = contMetrics
+		timestamps[podIdx] = api.TimeInfo{
+			Timestamp: earliestTS,
 			Window:    kubernetesCadvisorWindow,
```

Comment:
Now we can calculate real window based on difference in timestamps.

Reply:
I changed to a computation of the time between the scrapes.